Add Vitess example for sharded MySQL in Kubernetes.

enisoc opened the pull request:
This is a simplified version of the configs found in the Vitess repo:

https://github.com/youtube/vitess/tree/master/examples/kubernetes

Here we use a single script to start all of Vitess, since the aim is just to show the end result. The full tutorial on our site goes into much more detail on each step:

http://vitess.io/getting-started/

Commit:
Add Vitess example for sharded MySQL in Kubernetes.

This is a simplified version of the configs found in the Vitess repo:

https://github.com/youtube/vitess/tree/master/examples/kubernetes

Here we use a single script to start all of Vitess, since the aim is
just to show the end result. The full tutorial on our site goes into
much more detail on each step:

http://vitess.io/getting-started/

bprashanth commented:
@enisoc Thanks!
 
Having suffererd with mysql in a past life I feel this example is extremely valuable , hence the comments that follow. Feel free to tell me you don't have the time and I'll just do a normal code reivew.

The example as it stands is a pretty opaque box that focuses on vitess. People are very confused about mysql on kube in general, if we got this example right I can just point people at it for all questions along the lines of 'how do I shard/replicate/scale mysql on kubernetes'.

Some suggestions: 
* Break the vitess-up script into subsections. Eg:     
   1. Setup vttablet: A vttablet is a unit of scaling that consists of a mysql and vttablet process running in the same pod.     
  create vttablet pod: kubect create -f somepod    
  verify vttablet works: $ kvtctl listalltablets    
  create a table ...
  2. Start vtgate: vtgate routes queries to tablets...

you get the idea.
    
* Switch guestbook out for a much simpler test, even something manual like insert this row, check these tablets. Inserting simple tests at each step above will also suffice.

* Add a trouble shooting section
* Add an advanced section, where you hint at something like failure or resharding. 

Though somethings better than nothing, so i'll just take this if you're pegged for time. I'll do a more thorough code review after your reply.

bprashanth commented on examples/vitess/etcd-controller-template.yaml:
```diff
@@ -0,0 +1,55 @@
+  selector:
+    name: etcd
+    cell: {{cell}}
+  template:
```
please set resource limits on this and other pods/controllers so the scheduler can spread them efficiently: https://github.com/kubernetes/kubernetes/blob/release-1.0/docs/user-guide/compute-resources.md

bprashanth commented on examples/vitess/vtctld-pod.yaml:
```diff
@@ -0,0 +1,37 @@
+apiVersion: v1
+metadata:
+  name: vtctld
+  labels:
```
please append a label that allows me to kill all vitess related pods/rcs/svc with kubectl delete po,rc,svc -l app=vitess

bprashanth commented on examples/vitess/vtctld-service.yaml:
```diff
@@ -0,0 +1,14 @@
+  ports:
+    - port: 15000
+      targetPort: 15000
+      nodePort: 30000
```
do we have to hardcode 3000? other 2 are ok but nodePort needs to be available on all nodes in the cluster, so its a little awkward.

enisoc replied:
Is there a way to pick one automatically? Or do you mean just provide a way for the user to pick one and hand it to the script?

bprashanth replied:
if you don't specify nodePort (i.e type: NodePort but no nodePort setting in the svc) a random port should get picked for you

enisoc replied:
I tried to make this work, but the problem is opening the port in the firewall. Since (afaik) there's no automated way to do that across platforms (e.g. GCE vs AWS), I'm telling the user to do that manually and pointing them to some docs.

If I don't know the port in advance, `vitess-up.sh` can't do everything it needs without user intervention, which defeats the purpose. I understand you're interested in splitting it up in a later iteration, but for now I don't think making the port automatic is worth compromising the one-click goal of the current incarnation.

Let me know if you feel strongly about it. In that case, I would probably have to do some work to split the tutorial up to be closer to your ideal granularity before I can make the port automatic.

bprashanth commented:
LGTM sans nits. Leave it to your if you want to address the suggestions in this/a follow up pr/or just point them at the vitess docs, though I would prefer something less opaque.

bprashanth commented on examples/vitess/vtgate-up.sh:
```diff
@@ -0,0 +1,44 @@
+
+vtdataroot_volume='emptyDir: {}'
+if [ -n "$VTDATAROOT_VOLUME" ]; then
+  vtdataroot_volume="hostDir: {path: ${VTDATAROOT_VOLUME}}"
```
This won't work in v1 right? it's not called hostPath

bprashanth commented on examples/vitess/vttablet-pod-template.yaml:
```diff
@@ -0,0 +1,108 @@
+          -db-config-filtered-uname vt_filtered
+          -db-config-filtered-dbname vt_{{keyspace}}
+          -db-config-filtered-charset utf8
+          -bootstrap_archive mysql-db-dir_10.0.13-MariaDB.tbz" vitess
```
please explain what this tbz is somewhere

enisoc replied:
FYI: It's a snapshot of a mysql data dir similar to what you'd get from a regular mysql data dir bootstrap, but with some preconfigured users for Vitess. This is because Vitess is designed to assume responsibility for access control, and then all connections to actual mysqld are the same user, which allows for connection pooling.

I'll try to find a good place to add this as a comment.

enisoc commented:
Thanks for the review. I'm definitely up for contributing more if it'll help fill in the story for MySQL on Kubernetes, since our goal is to be the answer to that question.

We have a full tutorial that goes into many of the details you mention: http://vitess.io/getting-started/

The reason I didn't include all that here is because it would become a hefty document that shares a lot of content with our own, and then I would have to maintain two copies. So my plan was to put something here that:

1) makes using Vitess in Kubernetes look really easy, hence the single setup step.
2) omits details that may change over time, so I don't need to change them in two places.

If you're interested in having the full doc here, I can work out a way to either make them identical, or make the differences purely mechanical and automated by a transform script. Then I can iterate on a single doc that's posted in two places. Let me know what you think, or if you have any other ideas for how to handle it.

bprashanth commented on examples/vitess/vttablet-pod-template.yaml:
```diff
@@ -0,0 +1,108 @@
+  volumes:
+    - name: syslog
+      hostPath: {path: /dev/log}
+    - name: vtdataroot
```
for my own education, when this pod gets whacked and restarted and admin would have to run CopySchemaShard? if so, please document this somewhere as it is one of the first things that trips people when managing data via hostpath.

enisoc replied:
If a pod gets restarted with an empty hostpath, it's supposed to restore from the latest backup. We have a backup/restore system that works on network mounts, but our Kubernetes example doesn't have that configured right now. We plan to add it soon, probably by dealing directly with GCS/S3, since we don't see a good way to dynamically request access to a particular network mount from inside a pod (I believe it would have to already be mounted into every node?).

bprashanth replied:
Hmm, so basically this won't do what it says on the tin (transparently replicate/shard mysql data) because of limitations in the kubernetes api, which is ok, as long as it's surfaced somewhere. Just adding a limitations section, or a wishlist at the end/in your docs is fine too. 

The kubernetes api is very much a wip, @saad-ali might be able to comment

bprashanth commented:
Having people jump around to get context can backfire (we faced similar problems with gce x kubernetes documentation), but I like your getting started guide and agree that duplication is bad. We can get this in as is, I'll try it when i have some time, we can wait for more people to do so, and iterate.

enisoc commented:
@bprashanth I believe I've addressed your comments, with the exception of the pre-assigned nodePort as mentioned above. Please let me know what you think. Thanks!

enisoc commented:
@bprashanth ping :)

bprashanth commented:
@enisoc sorry, got sidetracked and completely forgot about this. Will review tomorrow.

bprashanth commented:
LGTM, will apply label once shippable passes. Closing and reopening to force a run, since the last one happened 21 days ago and that's too old. Sorry for the delay, once again.

enisoc commented:
@bprashanth It looks like Shippable failed due to a flaky test. Can we trigger a retry?

bprashanth commented:
Flake is https://github.com/kubernetes/kubernetes/issues/14284, retrying

bprashanth commented:
@k8s-bot ok to test

bgrant0607 commented on examples/vitess/README.md:
````diff
@@ -0,0 +1,133 @@
+to expose it on an external IP. It may take a minute to set up, but you should
+soon see the external IP show up under the internal one like this:
+
+```
````
Please add formatting: ```console

bgrant0607 commented on examples/vitess/etcd-controller-template.yaml:
```diff
@@ -0,0 +1,61 @@
+metadata:
+  name: etcd-{{cell}}
+  labels:
+    name: etcd
```
Suggest `component` rather than `name`
https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/config-best-practices.md

bgrant0607 commented on examples/vitess/etcd-controller-template.yaml:
```diff
@@ -0,0 +1,61 @@
+    app: vitess
+spec:
+  replicas: {{replicas}}
+  selector:
```
You want this selector to match pods that are not labeled `app: vitess`?

If you want the selector to match the pod template labels exactly, just omit it.

bgrant0607 commented on examples/vitess/vtctld-pod.yaml:
```diff
@@ -0,0 +1,42 @@
+kind: Pod
```
You don't need this to keep running?

https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/pods.md#durability-of-pods-or-lack-thereof

bgrant0607 commented:
This should also be squashed before merging:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md#7-fix-feedback-in-a-new-commit

enisoc commented:
@bgrant0607 I believe I've addressed your comments. Please take a look. Thanks.

bgrant0607 commented:
LGTM, thanks.

enisoc commented:
The Shippable failure appears to be a flake. Can we try again?

bgrant0607 commented:
cc @k8s-oncall
